interface conversion: checker.TypeData is *checker.UnionType, not *checker.TypeParameter

## Stack trace

```
panic: interface conversion: checker.TypeData is *checker.UnionType, not *checker.TypeParameter

goroutine 143 [running]:
github.com/microsoft/typescript-go/internal/checker.(*Type).AsTypeParameter(...)
	/home/vlk/.cache/paru/clone/typescript-go-git/src/typescript-go/internal/checker/types.go:588
github.com/microsoft/typescript-go/internal/checker.(*Checker).getConstraintFromTypeParameter(0xc069f99308?, 0xc07fdaf8c0?)
	/home/vlk/.cache/paru/clone/typescript-go-git/src/typescript-go/internal/checker/checker.go:16269 +0x1db
github.com/microsoft/typescript-go/internal/checker.(*Checker).getConstraintOfTypeParameter(0xc069f99308, 0xc07fdaf8c0)
	/home/vlk/.cache/paru/clone/typescript-go-git/src/typescript-go/internal/checker/checker.go:16258 +0x45
github.com/microsoft/typescript-go/internal/checker.(*Relater).structuredTypeRelatedToWorker(0xc01079fd40, 0xc07fd8ea80, 0xc07fd8ea10, 0x1, 0x0)
	/home/vlk/.cache/paru/clone/typescript-go-git/src/typescript-go/internal/checker/relater.go:3375 +0x39c5
github.com/microsoft/typescript-go/internal/checker.(*Relater).structuredTypeRelatedTo(0xc01079fd40, 0xc07fd8ea80, 0xc07fd8ea10, 0x1, 0x0)
	/home/vlk/.cache/paru/clone/typescript-go-git/src/typescript-go/internal/checker/relater.go:3140 +0x65
github.com/microsoft/typescript-go/internal/checker.(*Relater).recursiveTypeRelatedTo(0xc01079fd40, 0xc07fd8ea80, 0xc07fd8ea10, 0x1, 0x0, 0x3)
	/home/vlk/.cache/paru/clone/typescript-go-git/src/typescript-go/internal/checker/relater.go:3081 +0x785
github.com/microsoft/typescript-go/internal/checker.(*Relater).isRelatedToEx(0xc01079fd40, 0xc07fd8ea80, 0xc07fd8ea10, 0x3, 0x1, 0x12fef00, 0x0)
	/home/vlk/.cache/paru/clone/typescript-go-git/src/typescript-go/internal/checker/relater.go:2656 +0x9a5
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkTypeRelatedToEx(0xc069f99308, 0xc07fd8ea80, 0xc07fd8ea10, 0xc072c82c10, 0xc00071f240, 0x12fef00, 0x0)
	/home/vlk/.cache/paru/clone/typescript-go-git/src/typescript-go/internal/checker/relater.go:373 +0x173
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkTypeComparableTo(...)
	/home/vlk/.cache/paru/clone/typescript-go-git/src/typescript-go/internal/checker/relater.go:351
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkAssertionDeferred(0xc069f99308, 0xc00071f240)
	/home/vlk/.cache/paru/clone/typescript-go-git/src/typescript-go/internal/checker/checker.go:11758 +0xf7
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkDeferredNode(0xc069f99308, 0xd9ad60?)
	/home/vlk/.cache/paru/clone/typescript-go-git/src/typescript-go/internal/checker/checker.go:2329 +0x1a5
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkDeferredNodes-range1(...)
	/home/vlk/.cache/paru/clone/typescript-go-git/src/typescript-go/internal/checker/checker.go:2302
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkDeferredNodes.(*OrderedSet[...]).Keys.func1(...)
	/home/vlk/.cache/paru/clone/typescript-go-git/src/typescript-go/internal/collections/ordered_map.go:129
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkDeferredNodes(0xc069f99308, 0xc006ba2bf0?)
	/home/vlk/.cache/paru/clone/typescript-go-git/src/typescript-go/internal/checker/checker.go:2298 +0xc5
github.com/microsoft/typescript-go/internal/checker.(*Checker).checkSourceFile(0xc069f99308, {0xd96428, 0x133e820}, 0xc0031fc848)
	/home/vlk/.cache/paru/clone/typescript-go-git/src/typescript-go/internal/checker/checker.go:2105 +0x12b
github.com/microsoft/typescript-go/internal/checker.(*Checker).CheckSourceFile(0xc069f99308, {0xd96428, 0x133e820}, 0xc0031fc848)
	/home/vlk/.cache/paru/clone/typescript-go-git/src/typescript-go/internal/checker/checker.go:2093 +0x5e
github.com/microsoft/typescript-go/internal/compiler.(*Program).CheckSourceFiles.func1-range1(0xc0754e2701?)
	/home/vlk/.cache/paru/clone/typescript-go-git/src/typescript-go/internal/compiler/program.go:289 +0x54
github.com/microsoft/typescript-go/internal/compiler.(*checkerPool).Files.func1(0xc05a0e8480)
	/home/vlk/.cache/paru/clone/typescript-go-git/src/typescript-go/internal/compiler/checkerpool.go:82 +0x88
github.com/microsoft/typescript-go/internal/compiler.(*Program).CheckSourceFiles.func1()
	/home/vlk/.cache/paru/clone/typescript-go-git/src/typescript-go/internal/compiler/program.go:287 +0x118
github.com/microsoft/typescript-go/internal/core.(*parallelWorkGroup).Queue.func1()
	/home/vlk/.cache/paru/clone/typescript-go-git/src/typescript-go/internal/core/workgroup.go:39 +0x50
created by github.com/microsoft/typescript-go/internal/core.(*parallelWorkGroup).Queue in goroutine 1
	/home/vlk/.cache/paru/clone/typescript-go-git/src/typescript-go/internal/core/workgroup.go:37 +0x85
```

## Steps to reproduce

1. Launch Minecraft with the Kubejs and probejs mods installed, get a probejs dump
2. write some code
3. crash!

I can't really upload the actual ts file itself, but I can paste it here!

I believe that my function `RECIPE_REGISTRAR.utils.derive_object_or_tagged` might be causing some problems, if anyone knows how I can make it better just generally, let me know. I do have a TODO on it to simplify it after all...

<details>

```ts
/**
 * For those unfamiliar with a kubejs codebase:
 *
 * namespace Internal = global stuff related to the Java runtime
 * namespace Special = stuff autogenerated by probejs, basically what it finds are in the registries
 * namespace inttypes = A file full of gigantic number unions. I did it this way because this is probably
 *      more performant than using the recursive array trick
 * namespace tables = A bunch of tables autogenerated by build.sh from a text file. More string unions but also some arrays
 *
 * type Internal.CompoundTag_ = string | Internal.CompoundTag | { [x: string]: number | boolean | Internal.CompoundTag_; }
 * type Special.Item / Fluid / Block \ etc... = a gigantic union of strings
 * type Special.ItemTag / FluidTag / BlockTag / etc... = another gigantic union of strings,
 *      that CAN BE covariant with Special.Item / Fluid / Block or whatever
 *
 */

export namespace mcty {
    export type nbt_t = Internal.CompoundTag_

    export type item_t = Special.Item
    export type fluid_t = Special.Fluid
    export type block_t = Special.Block

    export type itemtag_t = Special.ItemTag
    export type fluidtag_t = Special.FluidTag
    export type blocktag_t = Special.BlockTag

    export type direction_t = "down" | "up" | "north" | "south" | "west" | "east"

    export type MoltenFluid = {
        fluid: mcty.fluid_t
        equivalent_tag?: mcty.fluidtag_t
        temperature: inttypes.celsius_t
    }

    export type ThingOrThingTag<T extends string, H extends string> = T | `#${H}`

    export type ItemOrHashTag = ThingOrThingTag<mcty.item_t, mcty.itemtag_t>
    export type FluidOrHashTag = ThingOrThingTag<mcty.fluid_t, mcty.fluidtag_t>
    export type BlockOrHashTag = ThingOrThingTag<mcty.block_t, mcty.blocktag_t>
}


/**
 * Internal types for this library
 * These are here to declutter code as well as keep track of which individual mods
 * use what kind of naming conventions, in a way that is more reliable than just `grep`ping.
 */
export namespace rcpt {

    export type I__item = { item: mcty.item_t }
    export type I__id = { id: I__item["item"] }

    export type I__tag = { tag: mcty.itemtag_t }

    export type I__item_tag = I__item | I__tag

    export type IC__count = { count: inttypes.stacksize_t }
    export type IC__Count = { Count: IC__count["count"] }
    export type IC__amount = { amount: IC__count["count"] }
    export type IC__Amount = { Amount: IC__count["count"] }

    export type F__fluid = { fluid: mcty.fluid_t }
    export type F__Fluid = { Fluid: F__fluid["fluid"] }
    export type F__FluidType = { FluidType: F__fluid["fluid"] }
    export type F__FluidName = { FluidName: F__fluid["fluid"] }

    export type F__tag = { tag: mcty.fluidtag_t }
    export type F__fluid_tag = { fluid_tag: F__tag["tag"] }
    export type F__FluidTag = { FluidTag: F__tag["tag"] }
    export type F__fluidTag = { fluidTag: F__tag["tag"] }

    export type FC__amount = { amount: inttypes.millibucket_t }
    export type FC__Amount = { Amount: FC__amount["amount"] }
    export type FC__millibuckets = { millibuckets: FC__amount["amount"] }
    export type FC__Millibuckets = { Millibuckets: FC__amount["amount"] }
    export type FC__MilliBuckets = { MilliBuckets: FC__amount["amount"] }

    export type C__nbt = { nbt: mcty.nbt_t }
    export type C__Nbt = { Nbt: C__nbt["nbt"] }

    export type C__chance = { chance: inttypes.float }
}

export type Optional<T extends { [key in string]: any }> = { [K in keyof T]?: T[K] }

export namespace recipe_json {
    /**
     * The base interface for all recipe types added by this library. You may specify it in functions
     * that accept recipes, and you can, of course, extend it with your own.
     */
    export interface RecipeKind<T extends Special.RecipeSerializer> {
        /** The type of the recipe. Always required */
        type: T
        /**
         * Forge conditions (whether or not to load the recipe), see the
         * [docs](https://docs.minecraftforge.net/en/latest/resources/server/conditional/#conditions)
         * for more information
         */
        conditions?: any
    }

    /**
     * A recipe kind that is impossible to instantiate if the required type is not supported.
     * Great for debugging mod removals.
     *
     * If you get weird type errors leading back to this type,
     * the tests in `test_ConditionalRecipeKind` probably failed.
     */
    export type ConditionalRecipeKind<T> = test_ConditionalRecipeKind.AreTestsPassing extends true
        ? test_ConditionalRecipeKind.ConditionalRecipeKindInner<T>
        : never

    namespace test_ConditionalRecipeKind {
        export interface Invalid<T> {
            user_provided_type: T
        }
        export type ConditionalRecipeKindInner<T> = T extends Special.RecipeSerializer ? RecipeKind<T> : Invalid<T>
        interface KnownGood extends ConditionalRecipeKindInner<"minecraft:blasting"> { }
        interface KnownBad extends ConditionalRecipeKindInner<"You're a BOYKISSER!"> { }
        type FirstTestPasses = KnownGood extends RecipeKind<infer T> ? true : false
        type SecondTestPasses = KnownBad extends RecipeKind<infer T> ? true : false

        export type AreTestsPassing = FirstTestPasses extends true
            ? SecondTestPasses extends false
            ? true
            : false /* \SecondTestPasses */
            : false /* \FirstTestPasses */
    }

    export namespace farmers_delight {
        export interface Cooking extends RecipeKind<"farmersdelight:cooking"> {
            cookingtime: inttypes.tick_t
            experience: inttypes.float
            ingredients: rcpt.I__item_tag | rcpt.I__item_tag[] | rcpt.I__item_tag[][]
            recipe_book_tab?: "drinks" | "meals" | "misc"
            result: rcpt.I__item & Optional<rcpt.IC__count>
            container?: rcpt.I__item_tag
        }

        export interface Cutting extends RecipeKind<"farmersdelight:cutting"> {
            ingredients: [rcpt.I__item_tag] & rcpt.I__item_tag[]
            result: Array<rcpt.I__item & Optional<rcpt.C__chance> & Optional<rcpt.IC__count>>
            tool: rcpt.I__item_tag | (rcpt.I__item_tag & {
                type: "farmersdelight:tool_action";
                action: "axe_dig" | "axe_strip" | "pickaxe_dig" | "shovel_dig"
            })
            sound?: Special.SoundEvent
        }
    }

    export namespace tconstruct {
        export type TconFluid = (rcpt.F__fluid | rcpt.F__tag) & rcpt.FC__amount
        type ntc_item = rcpt.I__item & rcpt.IC__count
        type i_entity = { type: Special.EntityType } | { tag: Special.EntityTypeTag }
        type CastIngredient = rcpt.I__item_tag & { type?: "tconstruct:material" | "tconstruct:block_tag" }

        /** One-fifth of a second! */
        export type TconSmelteryTicks = inttypes.uint
        export type ModifierLevel = inttypes.uint | { max: inttypes.uint }

        export type TconstructMaterial = ResourceLocation_
        export type TconstructModifier = ResourceLocation_
        export type TconstructMaterialItem = ResourceLocation_

        export type TconCastTag_t = `tconstruct:casts/${"single" | "multi"}_use/${tables.TconCastKindStr}`


        export interface AgeableSevering extends RecipeKind<"tconstruct:ageable_severing"> {
            adult_result: mcty.item_t | ntc_item
            entity: i_entity
            child_result?: mcty.item_t | ntc_item
        }

        export interface Alloy extends RecipeKind<"tconstruct:alloy"> {
            inputs: Array<TconFluid & { catalyst?: boolean }>
            result: TconFluid
            temperature: inttypes.celsius_t
        }

        export interface Filling extends RecipeKind<"tconstruct:basin_filling" | "tconstruct:table_filling"> {
            container: mcty.item_t
            fluid_amount: inttypes.millibucket_t
        }

        export interface Casting extends RecipeKind<"tconstruct:casting_basin" | "tconstruct:casting_table"> {
            cast?: CastIngredient
            cast_consumed?: boolean
            cooling_time: inttypes.tick_t
            fluid: TconFluid
            result: rcpt.I__item_tag & Optional<rcpt.C__nbt>
            switch_slots?: boolean
        }

        export interface CastingTablePotion extends RecipeKind<"tconstruct:casting_table_potion"> {
            bottle: rcpt.I__item_tag
            fluid: TconFluid
            result: mcty.item_t
            cooling_time?: inttypes.tick_t
        }

        interface InternalMelting<T extends Special.RecipeSerializer, F extends TconFluid = TconFluid> extends RecipeKind<T> {
            ingredient: rcpt.I__item_tag
            result: F
            temperature: inttypes.celsius_t
            time: TconSmelteryTicks
            byproducts?: F[]
        }
        export type DamageableMelting = InternalMelting<"tconstruct:damagable_melting", TconFluid & { unit_size: inttypes.millibucket_t }>

        export interface EntityMelting extends RecipeKind<"tconstruct:entity_melting"> {
            damage: inttypes.float
            entity: i_entity[] | i_entity | { types: Special.EntityType[] }
            result: TconFluid & Optional<rcpt.C__nbt>
        }

        interface Slots {
            abilities?: inttypes.uint8_t
            defense?: inttypes.uint8_t
            upgrades?: inttypes.uint8_t
        }
        interface InternalModifier<T extends Special.RecipeSerializer> extends RecipeKind<T> {
            allow_crystal: boolean
            result: TconstructModifier
            slots?: Slots
        }
        export interface IncrementalModifier extends InternalModifier<"tconstruct:incremental_modifier"> {
            amount_per_item: inttypes.uint
            input: rcpt.I__item_tag
            level?: ModifierLevel
            needed_per_level: inttypes.stacksize_nonzero_t
            tools: rcpt.I__item_tag | rcpt.I__item_tag[]
            leftover?: mcty.item_t
        }

        export interface Material extends RecipeKind<"tconstruct:material"> {
            ingredient: rcpt.I__item_tag
            leftover?: rcpt.I__item_tag & Optional<rcpt.IC__count>
            material: TconstructMaterial
            needed: number
            value: number
        }

        export interface MaterialFluid extends RecipeKind<"tconstruct:material_fluid"> {
            fluid: TconFluid
            output: TconstructMaterial
            temperature: inttypes.celsius_t
            input?: TconstructMaterial
        }

        export interface MaterialMelting extends RecipeKind<"tconstruct:material_melting"> {
            input: TconstructMaterial
            result: TconFluid
            temperature: inttypes.celsius_t
        }

        export type Melting = InternalMelting<"tconstruct:melting">

        export interface MeltingFuel extends RecipeKind<"tconstruct:melting_fuel"> {
            /**
             * TODO: Does this use smeltery ticks or game ticks?
             */
            duration: TconSmelteryTicks
            fluid: TconFluid
            rate: inttypes.float
            temperature: inttypes.celsius_t
        }

        export interface Modifier extends InternalModifier<"tconstruct:modifier"> {
            check_trait_level?: boolean
            inputs: Array<rcpt.I__item_tag & { amount_needed?: inttypes.stacksize_nonzero_t }>
            level?: ModifierLevel
            tools: rcpt.I__item_tag | rcpt.I__item_tag[]
        }

        type OreMeltingRate = "metal" | "gem"
        export interface OreMelting extends InternalMelting<"tconstruct:ore_melting"> {
            type: "tconstruct:ore_melting";
            byproducts?: Array<TconFluid & { rate?: OreMeltingRate }>
            rate: OreMeltingRate;
            result: TconFluid & { rate?: OreMeltingRate }
        }

        export interface Severing extends RecipeKind<"tconstruct:severing"> {
            entity: { type: Special.EntityType } | { types: Special.EntityType[] }
            result: mcty.item_t | ntc_item
        }

        export interface TableCastingMaterial extends RecipeKind<"tconstruct:table_casting_material"> {
            cast: { tag: TconCastTag_t }
            item_cost: inttypes.stacksize_nonzero_t
            result: TconstructMaterialItem
            cast_consumed?: boolean
        }

        export interface TableCastingDuplication extends RecipeKind<"tconstruct:table_duplication" | "tconstruct:basin_duplication"> {
            cast: CastIngredient
            cooling_time: inttypes.tick_t
            fluid: TconFluid
        }

    }

    export namespace thermal {
        interface Machine<T extends Special.RecipeSerializer> extends RecipeKind<T> {
            energy?: inttypes.forge_energy_t
        }
        interface XpMachine<T extends Special.RecipeSerializer> extends Machine<T> {
            experience?: inttypes.udecimal_t
        }

        interface Catalyst<T extends Special.RecipeSerializer> extends RecipeKind<T> {
            ingredient: rcpt.I__item_tag
            primary_mod: inttypes.float
            secondary_mod: inttypes.float
            energy_mod: inttypes.float
            min_chance: inttypes.udecimal_t
            use_chance: inttypes.udecimal_t
        }

        type nt_fluid = rcpt.F__fluid & rcpt.FC__amount
        type t_fluid = (rcpt.F__fluid | rcpt.F__fluid_tag) & rcpt.FC__amount
        type optlocked = { locked?: boolean }

        export interface Bottler extends XpMachine<"thermal:bottler"> {
            ingredients: Array<(rcpt.I__item_tag | t_fluid) & Optional<rcpt.C__nbt>>
            result: Array<rcpt.I__item & Optional<rcpt.C__nbt>>
        }

        export interface Centrifuge extends XpMachine<"thermal:centrifuge"> {
            ingredient: (rcpt.I__item_tag) & rcpt.IC__count
            result: Array<optlocked & Optional<rcpt.C__chance> & ((rcpt.I__item & Optional<rcpt.IC__count>) | nt_fluid)>
        }

        interface ChillerInternal extends Machine<"thermal:chiller"> {
            result: rcpt.I__item & Optional<rcpt.IC__count>
        }
        // interface Chiller1 extends ChillerInternal { ingredient: t_fluid }
        // interface Chiller2 extends ChillerInternal { ingredients: [t_fluid, rcpt.I__item_tag?] }
        export type Chiller = ChillerInternal & ({ ingredient: t_fluid } | { ingredients: [t_fluid, rcpt.I__item_tag?] })

        export interface CompressionFuel extends Machine<"thermal:compression_fuel"> {
            ingredient: t_fluid
            energy: inttypes.forge_energy_t
        }

        export interface Crucible extends Machine<"thermal:crucible"> {
            ingredient: rcpt.I__item_tag
            result: nt_fluid[]
            energy: inttypes.forge_energy_t
        }

        export interface Crystallizer extends Machine<"thermal:crystallizer"> {
            ingredients: Array<t_fluid | rcpt.I__item_tag>
            result: [rcpt.I__item]
        }

        export interface RedstoneFurnace extends RecipeKind<"thermal:furnace"> {
            ingredient: rcpt.I__item_tag & Optional<rcpt.IC__count>
            result: rcpt.I__item
            experience?: inttypes.udecimal_t
            energy_mod?: inttypes.forge_energy_t
        }

        export interface Insolator extends XpMachine<"thermal:insolator"> {
            ingredient: rcpt.I__item_tag
            result: Array<rcpt.I__item & Optional<rcpt.C__chance>>
            energy_mod?: inttypes.float
            water_mod?: inttypes.float
        }

        export type InsolatorCatalyst = Catalyst<"thermal:insolator_catalyst">

        export interface MagmaticFuel extends RecipeKind<"thermal:magmatic_fuel"> {
            ingredient: t_fluid
            energy: inttypes.forge_energy_t
        }

        export interface NumismaticFuel extends RecipeKind<"thermal:numismatic_fuel"> {
            ingredient: rcpt.I__item_tag
            energy: inttypes.forge_energy_t
        }

        interface PressInternal extends XpMachine<"thermal:press"> {
            result: Array<(rcpt.I__item & Optional<rcpt.IC__count>) | nt_fluid>
        }
        export type Press = PressInternal & ({ ingredients: Array<rcpt.I__item_tag & Optional<rcpt.IC__count>> } | { ingredient: rcpt.I__item_tag })

        export interface Pulverizer extends XpMachine<"thermal:pulverizer" | "thermal:pulverizer_recycle"> {
            ingredient: rcpt.I__item_tag & Optional<rcpt.IC__count>
            result: Array<rcpt.I__item & Optional<rcpt.C__chance> & Optional<rcpt.IC__count> & optlocked>
            energy_mod?: inttypes.float
        }

        export type PulverizerCatalyst = Catalyst<"thermal:pulverizer_catalyst">

        export interface Pyrolyzer extends XpMachine<"thermal:pyrolyzer"> {
            ingredient: rcpt.I__item_tag
            result: Array<(rcpt.I__item & Optional<rcpt.C__chance>) | nt_fluid>
        }

        export interface Refinery extends XpMachine<"thermal:refinery"> {
            ingredient: t_fluid
            result: Array<(rcpt.I__item & Optional<rcpt.C__chance>) | nt_fluid>
        }

        export interface Sawmill extends XpMachine<"thermal:sawmill"> {
            ingredient: rcpt.I__item_tag
            result: Array<rcpt.I__item & Optional<rcpt.C__chance> & Optional<rcpt.IC__count>>
        }

        interface AlloySmelterInner extends XpMachine<"thermal:smelter" | "thermal:smelter_recycle"> {
            result: Array<rcpt.I__item & Optional<rcpt.C__chance> & Optional<rcpt.IC__count> & optlocked>
            energy_mod?: inttypes.float
        }
        export type AlloySmelter = AlloySmelterInner & ({ ingredients: Array<rcpt.I__item_tag & Optional<rcpt.IC__count>> } | { ingredient: rcpt.I__item_tag })

        export type AlloySmelterCatalyst = Catalyst<"thermal:smelter_catalyst">

        type NaturalLeaves = {
            Name: mcty.block_t
            Properties: {
                persistent: "false"
            }
        }
        export interface Extractor extends ConditionalRecipeKind<"thermal:tree_extractor"> {
            trunk: {
                Name: mcty.block_t
                Properties?: mcty.nbt_t
            }
            leaves: NaturalLeaves[] | NaturalLeaves
            sapling: mcty.block_t
            min_height: inttypes.y_overworld_t
            max_height: inttypes.y_overworld_t
            min_leaves: inttypes.uint8_t
            max_leaves: inttypes.uint8_t
            result: nt_fluid
        }
    }

    export namespace xycraft {
        export namespace predicate {
            export type predicate_t = "xycraft_core:block_rule" | "xycraft_core:fluid_type_rule" | "xycraft_core:fluid_tag_rule" | "xycraft_core:block_tag_rule"

            export interface ExtractorPredicate<P extends predicate_t> {
                predicate_type: P
            }
            export interface Block extends ExtractorPredicate<"xycraft_core:block_rule"> {
                block: mcty.block_t
            }
            export interface BlockTag extends ExtractorPredicate<"xycraft_core:block_tag_rule"> {
                tag: mcty.blocktag_t
            }
            export interface Fluid extends ExtractorPredicate<"xycraft_core:fluid_type_rule"> {
                fluid_type: mcty.fluid_t
            }
            export interface FluidTag extends ExtractorPredicate<"xycraft_core:fluid_tag_rule"> {
                tag: mcty.fluidtag_t
            }

            export type AllPredicateTypes = predicate.Block | predicate.BlockTag | predicate.Fluid | predicate.FluidTag
        }

        export interface Extractor extends ConditionalRecipeKind<"xycraft_machines:extractor"> {
            adjacent: SizedArray.Maybe<SizedArray.UpToFour<predicate.AllPredicateTypes>>
            item_result?: rcpt.I__id & Optional<rcpt.IC__Count>
            target: predicate.AllPredicateTypes
            ticks: inttypes.tick_t
            valid_directions: SizedArray.UpToSix<mcty.direction_t>
            waterlogged_fluid?: "minecraft:water" | "minecraft:lava"
            catalyst?: predicate.AllPredicateTypes
            fluid_result?: rcpt.FC__Amount & rcpt.F__FluidName
        }
    }
}

function f(recipe: recipe_json.thermal.Bottler) {
    // let H: recipe.Bottler = {

    // }
}
type H = {
    /**
     * Calculates the temperature for a recipe based on the given temperature and factor
     * @param temperature  Required melting temperature
     * @param amount       Amount of relevant fluid
     * @return  Time for the recipe in celsius
     */
    calcMeltingTimeForAmount: (
        temperature: inttypes.celsius_t,
        amount: inttypes.millibucket_t
    ) => number
    /**
     * Calculates the cooling time for a recipe based on the amount and temperature
     * @param temperature  Temperature baseline in celsius
     * @param amount       Output amount
     * @return  Cooling time based on the given inputs
     */
    calcCoolingTime: (
        temperature: inttypes.celsius_t,
        amount: inttypes.millibucket_t
    ) => number
}
var RECIPE_REGISTRAR = {
    add_recipe<R extends recipe_json.RecipeKind<any>>(recipe_event: Internal.RecipesEventJS_, recipe: R) {
        recipe_event.custom(recipe)
    },
    utils: {
        /**
         * NOTE: This function confuses `tsgo` a lot.
         *
         * TODO: Figure out some way to not have to have 3 generics on this fn
         */
        derive_object_or_tagged<A, B, T extends mcty.ThingOrThingTag<A, B>>(thing: T): ({ thing: A, thing_tag?: never } | { thing?: never, thing_tag: B }) {
            if (thing.startsWith("#")) {
                return { thing_tag: thing as B }
            } else {
                return { thing: thing as A }
            }
        }
    },
    tconstruct: {
        utils: {
            /**
             * Calculates the temperature for a recipe based on the given temperature and factor.
             * base formula is temp^(.585), which will produce a time in 1/5th second increments.
             * @param temperature  Required melting temperature in Celsius
             * @param factor       Multiplier based on material type
             * @return  Time for the recipe in celsius
             */
            calcMeltingTime(temperature: inttypes.celsius_t, factor: number): recipe_json.tconstruct.TconSmelteryTicks {
                return JavaMath.round(
                    JavaMath.pow(temperature /*  + 300.0 */, 0.585) * factor
                ) as any
            },

            /**
            * Calculates a time factor for the given fluid amount
            * @param fluid_amount  Fluid amount
            * @return  Time factor
            */
            calcTimeFactor(fluid_amount: inttypes.millibucket_t): inttypes.float {
                return JavaMath.sqrt(fluid_amount / 90 /* ingot amount */)
            },

            /**
            * Calculates the temperature for a recipe based on the given temperature and factor
            * @param temperature  Required melting temperature
            * @param fluid_amount       Amount of relevant fluid
            * @return  Time for the recipe in celsius
            */
            calcMeltingTimeForAmount(temperature: inttypes.celsius_t, fluid_amount: inttypes.millibucket_t): recipe_json.tconstruct.TconSmelteryTicks {
                return RECIPE_REGISTRAR.tconstruct.utils.calcMeltingTime(
                    temperature,
                    RECIPE_REGISTRAR.tconstruct.utils.calcTimeFactor(fluid_amount)
                )
            },

            /**
             * Unlike the other functions here, this one was not copied verbatim from the Tinker's Construct src
             */
            calcMeltingTimeForFactorOrAmount(temperature: inttypes.celsius_t, fluid_amount?: inttypes.millibucket_t, factor?: number): recipe_json.tconstruct.TconSmelteryTicks {
                if (factor === undefined) {
                    if (fluid_amount === undefined) {
                        throw new TypeError("calcMeltingTimeForFactorOrAmount: factor or fluid_amount is required")
                    }

                    factor = RECIPE_REGISTRAR.tconstruct.utils.calcTimeFactor(fluid_amount)
                }
                return RECIPE_REGISTRAR.tconstruct.utils.calcMeltingTime(
                    temperature,
                    factor
                )
            },

            /**
            * Calculates the cooling time for a recipe based on the amount and temperature
            * @param temperature  Temperature baseline in celsius
            * @param fluid_amount       Output amount
            * @return  Cooling time based on the given inputs
            */
            calcCoolingTime(temperature: inttypes.celsius_t, fluid_amount: inttypes.millibucket_t): inttypes.tick_t {
                // This is what Tinker's does internally
                return RECIPE_REGISTRAR.tconstruct.utils.calcMeltingTimeForAmount(temperature, fluid_amount)
            },

            convert_fluid(fluid: mcty.MoltenFluid, fluid_amount: inttypes.millibucket_t): recipe_json.tconstruct.TconFluid {
                if (fluid.equivalent_tag !== undefined) {
                    return {
                        amount: fluid_amount,
                        tag: fluid.equivalent_tag,
                    }
                } else {
                    return {
                        amount: fluid_amount,
                        fluid: fluid.fluid
                    }
                }
            }
        },

        add_casting_recipe(recipe_event: Internal.RecipesEventJS_,
            kind: "casting_basin" | "casting_table",
            result: mcty.ItemOrHashTag,
            fluid: mcty.MoltenFluid,
            fluid_amount: inttypes.millibucket_t,
            factor?: number
        ) {
            // uncomment for a free tsgo error!

            // RECIPE_REGISTRAR.add_recipe<recipe_json.tconstruct.Casting>(recipe_event, {
            //     type: `tconstruct:${kind}`,
            //     fluid: RECIPE_REGISTRAR.tconstruct.utils.convert_fluid(fluid, fluid_amount),
            //     cooling_time: RECIPE_REGISTRAR.tconstruct.utils.calcCoolingTime(fluid.temperature, fluid_amount),
            //     result
            // })
        },
    }
}

```

</details>

--- a/internal/checker/relater.go
+++ b/internal/checker/relater.go
@@ -3365,15 +3365,8 @@ func (r *Relater) structuredTypeRelatedToWorker(source *Type, target *Type, repo
 		if r.relation == r.c.comparableRelation && source.flags&TypeFlagsTypeParameter != 0 {
 			// This is a carve-out in comparability to essentially forbid comparing a type parameter with another type parameter
 			// unless one extends the other. (Remember: comparability is mostly bidirectional!)
-			constraint := r.c.getConstraintOfTypeParameter(source)
-			if constraint != nil {
-				for constraint != nil && someType(constraint, func(c *Type) bool { return c.flags&TypeFlagsTypeParameter != 0 }) {
-					result = r.isRelatedTo(constraint, target, RecursionFlagsSource, false /*reportErrors*/)
-					if result != TernaryFalse {
-						return result
-					}
-					constraint = r.c.getConstraintOfTypeParameter(constraint)
-				}
+			if constraint := r.c.getConstraintOfTypeParameter(source); constraint != nil && someType(constraint, func(c *Type) bool { return c.flags&TypeFlagsTypeParameter != 0 }) {
+				return r.isRelatedTo(constraint, target, RecursionFlagsSource, false /*reportErrors*/)
 			}
 			return TernaryFalse
 		}

--- a/testdata/baselines/reference/compiler/typeParameterComparableRelation.errors.txt
+++ b/testdata/baselines/reference/compiler/typeParameterComparableRelation.errors.txt
@@ -1,0 +1,16 @@
+typeParameterComparableRelation.ts(6,5): error TS2352: Conversion of type 'T' to type 'A' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
+  'A' could be instantiated with an arbitrary type which could be unrelated to 'T'.
+
+
+==== typeParameterComparableRelation.ts (1 errors) ====
+    // https://github.com/microsoft/typescript-go/issues/1462
+    
+    type StringOrT<T> = T | string
+    
+    function func<A, B, T extends StringOrT<B>>(thing: T): void {
+        thing as A; // Error
+        ~~~~~~~~~~
+!!! error TS2352: Conversion of type 'T' to type 'A' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
+!!! error TS2352:   'A' could be instantiated with an arbitrary type which could be unrelated to 'T'.
+    }
+    

--- a/testdata/baselines/reference/compiler/typeParameterComparableRelation.symbols
+++ b/testdata/baselines/reference/compiler/typeParameterComparableRelation.symbols
@@ -1,0 +1,25 @@
+//// [tests/cases/compiler/typeParameterComparableRelation.ts] ////
+
+=== typeParameterComparableRelation.ts ===
+// https://github.com/microsoft/typescript-go/issues/1462
+
+type StringOrT<T> = T | string
+>StringOrT : Symbol(StringOrT, Decl(typeParameterComparableRelation.ts, 0, 0))
+>T : Symbol(T, Decl(typeParameterComparableRelation.ts, 2, 15))
+>T : Symbol(T, Decl(typeParameterComparableRelation.ts, 2, 15))
+
+function func<A, B, T extends StringOrT<B>>(thing: T): void {
+>func : Symbol(func, Decl(typeParameterComparableRelation.ts, 2, 30))
+>A : Symbol(A, Decl(typeParameterComparableRelation.ts, 4, 14))
+>B : Symbol(B, Decl(typeParameterComparableRelation.ts, 4, 16))
+>T : Symbol(T, Decl(typeParameterComparableRelation.ts, 4, 19))
+>StringOrT : Symbol(StringOrT, Decl(typeParameterComparableRelation.ts, 0, 0))
+>B : Symbol(B, Decl(typeParameterComparableRelation.ts, 4, 16))
+>thing : Symbol(thing, Decl(typeParameterComparableRelation.ts, 4, 44))
+>T : Symbol(T, Decl(typeParameterComparableRelation.ts, 4, 19))
+
+    thing as A; // Error
+>thing : Symbol(thing, Decl(typeParameterComparableRelation.ts, 4, 44))
+>A : Symbol(A, Decl(typeParameterComparableRelation.ts, 4, 14))
+}
+

--- a/testdata/baselines/reference/compiler/typeParameterComparableRelation.types
+++ b/testdata/baselines/reference/compiler/typeParameterComparableRelation.types
@@ -1,0 +1,17 @@
+//// [tests/cases/compiler/typeParameterComparableRelation.ts] ////
+
+=== typeParameterComparableRelation.ts ===
+// https://github.com/microsoft/typescript-go/issues/1462
+
+type StringOrT<T> = T | string
+>StringOrT : StringOrT<T>
+
+function func<A, B, T extends StringOrT<B>>(thing: T): void {
+>func : <A, B, T extends StringOrT<B>>(thing: T) => void
+>thing : T
+
+    thing as A; // Error
+>thing as A : A
+>thing : T
+}
+

--- a/testdata/tests/cases/compiler/typeParameterComparableRelation.ts
+++ b/testdata/tests/cases/compiler/typeParameterComparableRelation.ts
@@ -1,0 +1,10 @@
+// @strict: true
+// @noEmit: true
+
+// https://github.com/microsoft/typescript-go/issues/1462
+
+type StringOrT<T> = T | string
+
+function func<A, B, T extends StringOrT<B>>(thing: T): void {
+    thing as A; // Error
+}

